### PR TITLE
Spark compatibility: Add new function space()

### PR DESCRIPTION
**Pull request by KevinyhZou**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added a function "space()" which repeats a space as many times as specified

### CODE REVIEW

File: src/Functions/repeat.cpp [hunk 1]
```diff
@@ -243,17 +243,74 @@ class FunctionRepeat : public IFunction
                 return res;
             }
         }
-
         throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Illegal column {} of argument of function {}",
             arguments[0].column->getName(), getName());
     }
 };
 
+class FunctionSpace : public IFunction
```
Comment (ucasfl) [inline src/Functions/repeat.cpp:251]: I think it's better to put it into a seperate file like `space.cpp`.

File: src/Functions/repeat.cpp [hunk 2]
```diff
@@ -243,17 +243,74 @@ class FunctionRepeat : public IFunction
...
+        String getName() const override { return name; }
+        size_t getNumberOfArguments() const override { return 1; }
+
+        bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+        bool useDefaultImplementationForConstants() const override { return true; }
+
+        DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+        {
+            if (arguments.size() > 0 && !isUnsignedInteger(arguments[0]))
```
Comment (ucasfl) [inline src/Functions/repeat.cpp:273]:
```suggestion
            if (arguments.size() !=1 || !isUnsignedInteger(arguments[0]))
```

Comment (ucasfl) [reply src/Functions/repeat.cpp:273]: You just have one argument and it's must be integer.

File: src/Functions/repeat.cpp [hunk 3]
```diff
@@ -243,17 +243,74 @@ class FunctionRepeat : public IFunction
...
+            {
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                                "Second argument of function {} must be unsigned integer, got {}",
+                                name, arguments[0]->getName());
```
Comment (ucasfl) [inline src/Functions/repeat.cpp:277]:
```suggestion
                throw Exception(ErrorCodes::BAD_ARGUMENTS,
                                "Function {} must have one argument with unsigned integer type",
                                name);
```

File: docs/en/sql-reference/functions/string-functions.md [hunk 4]
```diff
@@ -1196,3 +1196,41 @@ Result:
 Same as concatWithSeparator, the difference is that you need to ensure that concatWithSeparator(sep, expr1, expr2, expr3...) → result is injective, it will be used for optimization of GROUP BY.
 
 The function is named “injective” if it always returns different result for different values of arguments. In other words: different arguments never yield identical result.
+
+## space
+
+Repeats a string as many times as specified and concatenates the replicated values as a single string.
```
Comment (ucasfl) [inline docs/en/sql-reference/functions/string-functions.md:1202]:
```suggestion
Repeats a space as many times as specified and concatenates the replicated values as a single string.
```

File: tests/queries/0_stateless/02674_space_function.reference [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+                                                       56
+                    20
+     5
+       7
+
+NULL
+NULL
+NULL
+NULL
```
Comment (ucasfl) [inline tests/queries/0_stateless/02674_space_function.reference:25]: Do you generate reference manully? Check https://clickhouse.com/docs/en/development/tests/.

Comment (KevinyhZou) [reply tests/queries/0_stateless/02674_space_function.reference:25]: yes,  I will look at this. And thanks for your reviews.

Comment (ucasfl) [reply tests/queries/0_stateless/02674_space_function.reference:25]: You can generate reference by `clickhouse-client -nm < test.sql > test.reference`

**Comment by ucasfl** [comment]
Also, you should add a simple document like this: https://github.com/ClickHouse/ClickHouse/pull/46631/files#diff-53f644c12e0dec84244b039bf2a2b5fe1508939830bcef7028dee1c57b2240e3R103-R105

**Comment by KevinyhZou** [comment]
> I think it's better not move `repeat.cpp` to `repeat.h`, but just move `RepeatImpl` into `repeat.h` and keep `repeat.cpp`.

OK

File: src/Functions/repeat.h [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+#include <Functions/IFunction.h>
+#include <Functions/castTypeToEither.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_COLUMN;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
```
Comment (ucasfl) [inline src/Functions/repeat.h:18]: There two ErrorCodes should be put into `repeat.cpp`.

Comment (KevinyhZou) [reply src/Functions/repeat.h]: done

File: src/Functions/repeat.h [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnVector.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+#include <Functions/castTypeToEither.h>
+
```
[the remark this answers is not in the document]
Comment (KevinyhZou) [reply src/Functions/repeat.h:5]: done

File: src/Functions/repeat.h [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    static void constStrVectorRepeat(
+        std::string_view copy_str,
+        ColumnString::Chars & res_data,
+        ColumnString::Offsets & res_offsets,
+        const PaddedPODArray<T> & col_num)
+    {
+        UInt64 data_size = 0;
+        res_offsets.resize(col_num.size());
+        UInt64 str_size = copy_str.size();
+        UInt64 col_size = col_num.size();
```
Comment (taiyang-li) [inline src/Functions/repeat.h:99]: UInt64 -> size_t

Comment (KevinyhZou) [reply src/Functions/repeat.h]: done

**Review by vdimir — COMMENTED** [review]
Thank you for your PR!

The introduction of a new function may not be necessary, the desired functionality can be achieved using `repeat` function:

```
:) CREATE FUNCTION space AS (n) -> repeat(' ', n);

Ok.

0 rows in set. Elapsed: 0.009 sec. 


:) select '"' || space(number) || '"' from numbers(10);

┌─concat('"', repeat(' ', number), '"')─┐
│ ""                                    │
│ " "                                   │
│ "  "                                  │
│ "   "                                 │
│ "    "                                │
│ "     "                               │
│ "      "                              │
│ "       "                             │
│ "        "                            │
│ "         "                           │
└───────────────────────────────────────┘

10 rows in set. Elapsed: 0.010 sec. 

```

What are your thoughts on the possibility of selecting this solution to avoid redundancy in the codebase?

**Comment by KevinyhZou** [comment]
> Thank you for your PR!
> 
> The introduction of a new function may not be necessary, the desired functionality can be achieved using `repeat` function:
> 
> ```
> :) CREATE FUNCTION space AS (n) -> repeat(' ', n);
> 
> Ok.
> 
> 0 rows in set. Elapsed: 0.009 sec. 
> 
> 
> :) select '"' || space(number) || '"' from numbers(10);
> 
> ┌─concat('"', repeat(' ', number), '"')─┐
> │ ""                                    │
> │ " "                                   │
> │ "  "                                  │
> │ "   "                                 │
> │ "    "                                │
> │ "     "                               │
> │ "      "                              │
> │ "       "                             │
> │ "        "                            │
> │ "         "                           │
> └───────────────────────────────────────┘
> 
> 10 rows in set. Elapsed: 0.010 sec. 
> ```
> 
> What are your thoughts on the possibility of selecting this solution to avoid redundancy in the codebase?

Thanks for your advice.  And I'll try to transform space function to repeat in the upper layer, it should be ok as you say.

**Comment by rschu1ze** [comment]
> > The introduction of a new function may not be necessary, the desired functionality can be achieved using `repeat` function:
> > What are your thoughts on the possibility of selecting this solution to avoid redundancy in the codebase?
> 
> Thanks for your advice. And I'll try to transform space function to repeat in the upper layer, it should be ok as you say.

I agree that a high-level transform would conceptually be nicer, even if this PR currently copypastes only ca. 100 LOC. A good starting point for that would be Parsers/ExpressionListParsers.cpp, function `getFunctionLayer()` (at least as far as I see, maybe there are even better places).

**Comment by alexey-milovidov** [comment]
@rschu1ze `repeat` does `memcpy` while `space` can do `memset`, and it can justify a standalone implementation.

**Comment by rschu1ze** [comment]
Continued + reimplemented in #50103.

@KevinyhZou Thanks anyways!

**Pull request closed without merging**